### PR TITLE
fix(NumberInput): allow user to back out number to type input

### LINKED ISSUE #8127: Bug - Number Input - Cannot clear value to specify a new value
**Describe the problem**
We are using a number input where we have a min value of 1. If the user wants to specify a value like 20, the user isn't able to easily clear the current value and specify a new value. E.g. inserting the cursor in the input and deleting the current value will result in resetting the value to the minimum value, i.e. 1. This is an odd user interaction and not consistent with how many users expect to specify values within an input. 

**How do you reproduce the problem?**
Specify a minimum other than 0 in the [PF example where min values are set](https://www.patternfly.org/v4/components/number-input#with-unit-and-thresholds). Place the cursor after the value in the input. Hit Delete. The min value still displays. 

**Expected behavior**

- Allow the user to clear the value using the Delete key. This means that they will see no value displayed in the input. 
- If the user leaves the input without typing a new value, then on blur, set the value to the min value of 1
- If the user specifies a value less than 1 in the input, then on blur, set the value to the min value of 1

**Is this issue blocking you?**
We currently have a bug open for this issue that was raised by QE. It is not considered critical. The workaround is to highlight the value and type the new value, but this isn't an interaction pattern that is common or obvious for users. 

**What is your product and what release date are you targeting?**
Kafka

**Any other information?**

### PR DESCRIPTION
**Pull request by kmcfaul**
**What**: Closes #8127

Allows the number input to be empty to let a user type in a number to the input.

I did have to update a unit test which is a bit of an edge case, but if the user intentionally passes `null` to `value` then there will be a console error (because an input should not be null). However in the updated example the input is fine when you clear the number (value = '') out to type something in, and the blur handler will bound the input to the min or max values once you move focus away from the input.

### CODE REVIEW

File: packages/react-integration/demo-app-ts/src/components/demos/NumberInputDemo/NumberInputDemo.tsx [hunk 1]
```diff
@@ -49,7 +49,7 @@ export class NumberInputDemo extends Component<NumberInputDemoState> {
 
   onPlus2 = () => {
     this.setState({
-      value2: this.state.value2 + 1
+      value2: this.state.value2 + 1 || 1
```
Comment (gitdallas) [inline packages/react-integration/demo-app-ts/src/components/demos/NumberInputDemo/NumberInputDemo.tsx:52]: This would make -1 skip 0 and go to 1. I wonder if something like `(this.state.value2 || 0) + 1` would be better.
Might not matter too much as this is for cypress.

File: packages/react-core/src/components/NumberInput/examples/NumberInputUnitThreshold.tsx [hunk 2]
```diff
@@ -24,8 +24,17 @@ export const NumberInputUnitThreshold: React.FunctionComponent = () => {
 
   const onChange = (event: React.FormEvent<HTMLInputElement>) => {
     const target = event.target as HTMLInputElement;
-    const newValue = normalizeBetween(isNaN(+target.value) ? 0 : Number(target.value), minValue, maxValue);
-    setValue(newValue);
+    setValue(+target.value);
```
Comment (gitdallas) [inline packages/react-core/src/components/NumberInput/examples/NumberInputUnitThreshold.tsx:27]: The issues asks "Allow the user to clear the value using the Delete key. This means that they will see no value displayed in the input."

The "+" here wouldn't allow for that. As this is the example you are working with for this change, maybe this could be something like:
```
const value = (event.target as HTMLInputElement).value;
setValue(value === "" ? value : +value);
```

Comment (gitdallas) [reply packages/react-core/src/components/NumberInput/examples/NumberInputUnitThreshold.tsx:27]: Should the value type (NumberInput.tsx:13) also include empty string?

**Review by gitdallas — CHANGES_REQUESTED** [review]

File: packages/react-integration/demo-app-ts/src/components/demos/NumberInputDemo/NumberInputDemo.tsx (hunk 1 above)
Comment (kmcfaul) [reply packages/react-integration/demo-app-ts/src/components/demos/NumberInputDemo/NumberInputDemo.tsx:52]: Yeah I can update it as that is clearer. I considered just removing the test because passing null doesn't behave as the test expects anymore (because it used to internally snap null to 0 but that prevented the input from ever being cleared for typing).

**Review by jenny-s51 — COMMENTED** [review]
This is awesome @kmcfaul ! Works as expected and will also close #7659  🎉

Since this will be a common use case, it might be a good idea to update the existing examples (or add a separate example) to have `value={null}` set, since it won't be immediately clear to consumers how they can implement this behavior. WDYT?

**Comment by kmcfaul** [comment]
@jenny-s51 I added a blurb to the threshold example which I did update to show the behavior.

File: packages/react-core/src/components/NumberInput/examples/NumberInputCustomStep.tsx [hunk 3]
```diff
@@ -2,16 +2,16 @@ import React from 'react';
...
   const stepper = (stepValue: number) => {
-    setValue(value + stepValue);
+    setValue((value || 0) + stepValue);
   };
 
   const onChange = (event: React.FormEvent<HTMLInputElement>) => {
-    const target = event.target as HTMLInputElement;
-    setValue(Number(target.value));
+    const value = (event.target as HTMLInputElement).value;
+    setValue(value === '' ? value : +value);
```
Comment (jenny-s51) [inline packages/react-core/src/components/NumberInput/examples/NumberInputCustomStep.tsx:14]: Neat syntax update!!

**Review by jenny-s51 — APPROVED** [review]
LGTM @kmcfaul !

File: packages/react-integration/demo-app-ts/src/components/demos/NumberInputDemo/NumberInputDemo.tsx [hunk 4]
```diff
@@ -14,7 +14,7 @@ export class NumberInputDemo extends Component<NumberInputDemoState> {
 
   state: NumberInputDemoState = {
     value: 0,
-    value2: null
+    value2: undefined
```
Comment (gitdallas) [inline packages/react-integration/demo-app-ts/src/components/demos/NumberInputDemo/NumberInputDemo.tsx:17]: Maybe the test name should be changed at the end of numberInput.spec.ts as it is no longer testing null, and/or a null example added.

Might also be a good idea to cypress test the new functionality with:

- a backspace to create a value of "" and a blur to populate value
- typing beyond min/max and blur to see it snap to min/max

**Review by gitdallas — CHANGES_REQUESTED** [review]

**Review by gitdallas — APPROVED** [review]

**Review by mmenestr — APPROVED** [review]

File: packages/react-core/src/components/NumberInput/NumberInput.tsx [hunk 5]
```diff
@@ -10,7 +10,7 @@ import { TextInput } from '../TextInput';
 
 export interface NumberInputProps extends React.HTMLProps<HTMLDivElement> {
   /** Value of the number input */
-  value?: number | null;
+  value?: number | '';
```
[the remark this answers is not in the document]
Comment (kmcfaul) [reply packages/react-core/src/components/NumberInput/NumberInput.tsx:13]: Passing `null` will still work as long as the user accounts for it in the plus/minus handlers. The only difference is that null previously would have been set to `0` in the component. I'm not sure why it was part of the type declaration.

I can probably add it back in and re-add that check only for a null value if that works out better.

[the remark this answers is not in the document]
Comment (kmcfaul) [reply packages/react-core/src/components/NumberInput/NumberInput.tsx:13]: The user will have to account for non-number values in the plus/minus handlers to handle the empty string value as well (but this is the new feature). So our current examples still work with null.

[the remark this answers is not in the document]
Comment (tlabaj) [reply packages/react-core/src/components/NumberInput/NumberInput.tsx:13]: As long as null still work it should be fine.
